### PR TITLE
Add checks for pending Qt threads and timers in tests

### PR DESCRIPTION
**Pull request by Czaki**
# Description

During the maintenance of napari, I observed multiple situations where the test crashed because Qt was trying to access memory already released. All such identified events were connected with some timers/threads that were not stopped before the test ended. Then test objects were cleaned, but Qt still references such objects. 

In this PR, I have added a mechanism for early detection of such problems and failing problematic tests, not a random one. 

I tried to find all tests with such problems, but it may happen that not all were watched as it depends on many things we do not control, like the process scheduler that depends on the operating system. 

## Type of change

- [x] Refactor to avoid problems in the future. 

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

**Comment by brisvag** [comment]
Nice work; some more comments in the code would make it easier to follow! But in general I think I get the gist, and the strategy makes sense to me. On zulip you mentioned this is mostly taking care of `QTimer` issues; why is that the case, and what does that mean in terms of how to fix the now failing tests?

I'm not familiar with the qt internals enought to judge whether you used the right/wrong methods, so I'll defer to others on this matter.

**Comment by Czaki** [comment]
> On zulip you mentioned this is mostly taking care of `QTimer` issues; why is that the case, and what does that mean in terms of how to fix the now failing tests?

This means that for the tests that fails locally on my computer (and this CI that I test) the reason for the failure is that we start QTimer in the test and do not wait on its ends with ending the test. 
Based on two tests that I check the solution is to monkeypatch start function and disable starting a timer in this test as the test result does not depends on the Timer event result. 

If some time-triggered output is required then proper wait should be added at the end of the test

File: napari/conftest.py
Comment (andy-sweet) [inline napari/conftest.py]: [`QThread.wait`](https://doc.qt.io/qt-5/qthread.html#wait) was only added in Qt 5.15 - is it possible to bump the minimum version required in setup.cfg? Maybe for tests only?

File: napari/conftest.py
Comment (andy-sweet) [inline napari/conftest.py]: Not sure what `dkt` means here. Consider renaming this to `started_threads`?

File: napari/conftest.py
Comment (andy-sweet) [inline napari/conftest.py]: Consider renaming to `dangling_qthreads` to clarify this only captures `QThreads` (e.g. and not threads started with Python's `threading`).

It might be worth add `q` or `qt` to all of the fixtures defined here to clarify that these are qt related.

File: napari/conftest.py
Comment (andy-sweet) [inline napari/conftest.py]: Is this intended to cover all the [`QTimer::singleShot` static convenience methods right](https://doc.qt.io/qt-5/qtimer.html#static-public-members)? Seems like there a few missing. Or are some of these dropped in pyqt/pyside?

File: napari/conftest.py
Comment (andy-sweet) [inline napari/conftest.py:583]: Based on the [docs of `QThreadPool`](https://doc.qt.io/qt-5/qthreadpool.html#details), there should only be one pool per qt app. Should we expect more than one app to be running at once during setup/teardown of a test?

File: napari/conftest.py
Comment (andy-sweet) [inline napari/conftest.py]: Consider adding a descriptive adjective to the assertion message (e.g. `unfinished thread ...` or `dangling thread ...`). Same comment for other similar assertion messages in this PR.

**Review by andy-sweet — COMMENTED** [review]
In general, I'm supportive of an effort like this. My comments are mostly minor suggestions.

The main difficulty is that we can only patch python code, so we have to cover every python API point that could create any of these things. For example, that might be an issue for `QTimer::singleShot`. I guess that's unavoidable though.

My other big question is do we know when the setup/teardown code defined here will be called with respect to other setup/teardown code?

File: napari/conftest.py
Comment (Czaki) [reply napari/conftest.py]: I change it to `qtbot.wait_until`

File: napari/conftest.py
Comment (Czaki) [reply napari/conftest.py]: dkt ≈ dict

File: napari/conftest.py
Comment (Czaki) [reply napari/conftest.py]: changed

File: napari/conftest.py
Comment (Czaki) [reply napari/conftest.py]: It looks like i need to sit and fix it

File: napari/conftest.py
Comment (Czaki) [reply napari/conftest.py:583]: No. The docs mean that there will be one global `QThreadPoll`. But You still could create the next one.

File: napari/conftest.py
Comment (Czaki) [reply napari/conftest.py]: Could you check the current one?

File: napari/conftest.py
Comment (andy-sweet) [reply napari/conftest.py:583]: Ah, true. This seems fine then.

File: napari/conftest.py
Comment (andy-sweet) [reply napari/conftest.py]: Looks good now.

File: napari/conftest.py
Comment (andy-sweet) [inline napari/conftest.py]: I'm not sure what using `qtbot.waitUntil` as a context manager means. [The docs](https://pytest-qt.readthedocs.io/en/latest/reference.html#pytestqt.qtbot.QtBot.waitUntil) just show plain calls to it. So maybe you can just call it before/after `quit`?

File: napari/conftest.py
Comment (andy-sweet) [inline napari/conftest.py]:
```suggestion
        "If this test does not require a QTimer to pass you could monkeypatch it out. "
```

File: napari/conftest.py
Comment (andy-sweet) [inline napari/conftest.py]:
```suggestion
        "If this test does not require a QPropertyAnimation to pass you could monkeypatch it out. "
```

File: napari/conftest.py
Comment (andy-sweet) [reply napari/conftest.py]: Love the new messages! Just made a couple of typo suggestions.

File: napari/conftest.py
Comment (Czaki) [inline napari/conftest.py]:
```suggestion 
        thread.quit()
        qtbot.waitUntil(thread.isFinished, timeout=2000)

```

File: napari/conftest.py
Comment (Czaki) [reply napari/conftest.py]: True

File: napari/conftest.py
Comment (Czaki) [reply napari/conftest.py]: Merged. Thanks

File: napari/_tests/test_conftest_fixtures.py
Comment (andy-sweet) [inline napari/_tests/test_conftest_fixtures.py]: 1 second is a long time for a unit test. Currently, we only execute this sleep twice and we're unlikely to expand this test much in the future, so it's not a big problem.

But ideally I'd like to see use of locks here to both speed these test up and explicitly control concurrent execution (i.e. rather than rely on timing). To do that, you could define a Python lock as an attribute of `_TestThread` then acquire that in `run` and in the part of the calling test where you want to prevent `run` from finishing (e.g. for `test_qthread_running` you would add a `with t.lock:` context manager for `t.start()` and `assert t.isRunning()`).

Comment (Czaki) [reply napari/_tests/test_conftest_fixtures.py]: I rewrite in on `mutex` It looks better now?

Comment (andy-sweet) [reply napari/_tests/test_conftest_fixtures.py]: Looks great!

Comment (andy-sweet) [reply napari/_tests/test_conftest_fixtures.py]: My only small concern is what happens when any lines in between `lock` and `unlock` fail either now or in the future (i.e. if that happens, is it possible for the test to wait forever in that case? Or is there something else that protects us?

If not, I'd prefer a scoped mutex/lock (e.g. with `QMutexLocker`) or asserting the return value of `QMutex.tryLock` with a timeout in `TestThread.run`

File: napari/_tests/test_conftest_fixtures.py
Comment (andy-sweet) [inline napari/_tests/test_conftest_fixtures.py:40]: I assume this is testing that threads still run when timers are disabled? If so, I'd suggest either adding an implementation comment to the code to explain this, or alternatively defining a separate test for this.

**Comment by andy-sweet** [comment]
I assume this will become ready to review once you've found a way to make the tests pass? Happy to take another look then.

File: napari/_tests/test_conftest_fixtures.py
Comment (Czaki) [reply napari/_tests/test_conftest_fixtures.py:40]: I have added a comment explaining the second part of the test. Did you think that it should be splitted?

**Comment by Czaki** [comment]
> I assume this will become ready to review once you've found a way to make the tests pass? Happy to take another look then.

Yes. But I asked for an earlier review of the core mechanism as the selected way of implementing fixtures may impact on fixing tests.

File: napari/_tests/test_conftest_fixtures.py
Comment (andy-sweet) [reply napari/_tests/test_conftest_fixtures.py:40]: The comment is great, so no need to split.

**Comment by Czaki** [comment]
I have fixed the errors that I found in a current test run. But it is possible that in the future, it will fail on another test because of different scheduler work. But from my perspective, it is good as it will highlight errors in a place where it is.

**Review by brisvag — COMMENTED** [review]
Looks good to me on a high level; I'm not *super* comfortable with approving the deep qt stuff as I'm not very familiar with it, but I think we should get this in asap, so I'm trusting @Czaki and @andy-sweet for the details.

Why the changes to the manifest, though?

File: napari/_qt/widgets/qt_dims_slider.py
Comment (andy-sweet) [inline napari/_qt/widgets/qt_dims_slider.py:608]: I think a single-shot `QTimer` will only fire once no matter how many time you call `start`:

https://doc.qt.io/qt-5/qtimer.html#singleShot-prop
https://doc.qt.io/qt-5/qtimer.html#start

After pulling down this branch, playback no longer works - it just advances once.

So either this should not be a single-shot timer, or we should go back to the static call to `QTimer::singleShot`.

File: napari/_qt/qt_main_window.py
Comment (andy-sweet) [inline napari/_qt/qt_main_window.py]: Optional: consider a rename to clarify the goal of this helper method - e.g. `_throttle_cursor_to_status_connection`.

File: napari/_tests/test_viewer.py
Comment (andy-sweet) [inline napari/_tests/test_viewer.py:173]: Add an implementation/explanatory comment to clarify why we need this, since it is not immediately obvious in this context. By following the code, I think it's a proxy for detecting that the animation has finished - right? And we can't wait for the `QAbstractAnimation::finished` signal since we don't know if the animation is still accessible, as it is removed?

**Review by andy-sweet — CHANGES_REQUESTED** [review]
Mostly looks good, as if I understood correctly, there aren't many functional changes here - the effect of this PR is mostly to patch qthread(pools), qanimations, and qtimers in tests to detect when they are stopped.

However, one functionality this breaks playback over a dimension slider. Slightly worrying that no existing tests fail.

**Comment by andy-sweet** [comment]
> Why the changes to the manifest, though?

Same question from me!

File: napari/_qt/widgets/qt_dims_slider.py
Comment (andy-sweet) [reply napari/_qt/widgets/qt_dims_slider.py:608]: Are we still able to use `QTimer::singleShot` with the qtimer patching in test?

Comment (Czaki) [reply napari/_qt/widgets/qt_dims_slider.py:608]: > Are we still able to use `QTimer::singleShot` with the qtimer patching in test?

Then we need to find another way to detect that there is no pending timer before test end, 

But this looks like a threading problem. In my console, I have:

```
WARNING: QObject::startTimer: Timers cannot be started from another thread
23:18:59 WARNING QObject::startTimer: Timers cannot be started from another thread
```

Comment (andy-sweet) [reply napari/_qt/widgets/qt_dims_slider.py:608]: Yeah, I have that in my console too. I suspect `QTimer::singleShot` bypasses that by creating a single-shot `QTimer` on the main thread, or just enqueuing the appropriate event?

Comment (Czaki) [reply napari/_qt/widgets/qt_dims_slider.py:608]: Now it working

File: napari/_qt/qt_main_window.py
Comment (Czaki) [reply napari/_qt/qt_main_window.py]: changed

File: napari/_tests/test_viewer.py
Comment (Czaki) [reply napari/_tests/test_viewer.py:173]: could you check the comment text?

**Comment by Czaki** [comment]
> > Why the changes to the manifest, though?
> 
> Same question from me!

cleaning after the previous PR to remove some warnings on the build source tarball. I could remove it and add another PR.

File: napari/_tests/test_viewer.py
Comment (andy-sweet) [reply napari/_tests/test_viewer.py:173]: looks great - thanks!

**Comment by brisvag** [comment]
>  I could remove it and add another PR.

Yeah maybe better for clarity! Thanks :)

File: napari/_qt/widgets/qt_dims_slider.py
Comment (andy-sweet) [reply napari/_qt/widgets/qt_dims_slider.py:608]: Retested and it's working now. So I guess you can start a single-shot timer more than once - good to know!

File: napari/_qt/widgets/qt_dims_slider.py
Comment (andy-sweet) [inline napari/_qt/widgets/qt_dims_slider.py:636]: I got some warnings when closing napari: 

```
WARNING: QObject::killTimer: Timers cannot be stopped from another thread
11:00:36 WARNING QObject::killTimer: Timers cannot be stopped from another thread
WARNING: QObject::~QObject: Timers cannot be stopped from another thread
11:00:36 WARNING QObject::~QObject: Timers cannot be stopped from another thread
```

That might be related to the test failures?

Maybe we need to pass this as a parent to the `QTimer`?

Comment (Czaki) [reply napari/_qt/widgets/qt_dims_slider.py:636]: I have found that I terminate the worker and thread in the same place. It means that the termination event is not processed as the thread is ended. 

I solve it by move the worker again to the main thread and then terminating.

Comment (andy-sweet) [reply napari/_qt/widgets/qt_dims_slider.py:636]: That fixed the tests, but I still get the error message when using playback on a 3D image (displayed in 2D), then closing napari.

https://user-images.githubusercontent.com/2608297/206309081-c55a9545-44fd-4ca6-be89-49304c9fddef.mp4

Comment (Czaki) [reply napari/_qt/widgets/qt_dims_slider.py:636]: Ah. I fix only showing error when user click stop button (it happens sometimes). I will fix it also.

Comment (Czaki) [reply napari/_qt/widgets/qt_dims_slider.py:636]: @andy-sweet could you check now?

Comment (andy-sweet) [reply napari/_qt/widgets/qt_dims_slider.py:636]: Works now! I have some minor (hopefully!) follow-up questions on the changes that made this work. Thanks again for your work on this and for being patient with my questions!

File: napari/_qt/qt_main_window.py
Comment (andy-sweet) [inline napari/_qt/qt_main_window.py:440]: Is there a reason we can't rely on `QtViewer.closeEvent` to call `QtDims.stop`? It's not ideal that we have to call this line in two different places to handle the two different close paths in `QtMainWindow`.

Comment (Czaki) [reply napari/_qt/qt_main_window.py:440]: Because `QtViewer.closeEvent` is not called (I add print and nothing was printed). I do not understand why.

**Review by andy-sweet — APPROVED** [review]
Looks good now. My only ask is to add a PR description about why this change was made.

**Comment by Czaki** [comment]
> Looks good now. My only ask is to add a PR description about why this change was made.

@andy-sweet Could you check the current description?

**Comment by andy-sweet** [comment]
> @andy-sweet Could you check the current description?

Looks great - thanks! Feel free to merge this when you're ready.

**Comment by Czaki** [comment]
> Looks great - thanks! Feel free to merge this when you're ready.

Just hit on the problem with not terminating the timer. Updated code because of this.

File: napari/utils/_testsupport.py
Comment (lucyleeow) [inline napari/utils/_testsupport.py:303]: @Czaki could you explain the reason for this change?

I am doing some clean up of typos here and see that the docstring for `strict_qt` still mentions 'raise'? I don't think there is a way to raise just a warning any more. Just wanted to double check this before I update the docstring to new behaviour. Thanks

Comment (Czaki) [reply napari/utils/_testsupport.py:303]: I do not remember. Maybe this change should be reverted.

Comment (lucyleeow) [reply napari/utils/_testsupport.py:303]: You would know better than me! Could it be left over from when you were testing/working on the PR?

**Pull request merged**
